From 2408daead27500a3bf6c7cd73c2de5e741ad4e2c Mon Sep 17 00:00:00 2001
From: Aren Siekmeier <aren.siekmeier@gmail.com>
Date: Mon, 19 Apr 2021 18:08:28 +0900
Subject: [PATCH] patch for fairseq v0.10.2 to apply in fairseq venv
 site-packages - this version produced 2021-04-19 srconly and tgtonly
 corrected results

apply patch in venv/lib/python/site-packages directory with
$ patch -p1 < patchfile

(generate patchfile from this checked out commit with)
$ git format-patch v0.10.2
---
 fairseq/data/language_pair_dataset.py | 2 +-
 fairseq/tasks/translation.py          | 6 +++++-
 2 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/fairseq/data/language_pair_dataset.py b/fairseq/data/language_pair_dataset.py
index 62e7109b..b50c3223 100644
--- a/fairseq/data/language_pair_dataset.py
+++ b/fairseq/data/language_pair_dataset.py
@@ -230,7 +230,7 @@ class LanguagePairDataset(FairseqDataset):
         if tgt_dict is not None:
             assert src_dict.pad() == tgt_dict.pad()
             assert src_dict.eos() == tgt_dict.eos()
-            assert src_dict.unk() == tgt_dict.unk()
+        #     assert src_dict.unk() == tgt_dict.unk()
         if tgt is not None:
             assert len(src) == len(
                 tgt
diff --git a/fairseq/tasks/translation.py b/fairseq/tasks/translation.py
index 79007a6d..e0b07148 100644
--- a/fairseq/tasks/translation.py
+++ b/fairseq/tasks/translation.py
@@ -363,7 +363,11 @@ class TranslationTask(LegacyFairseqTask):
         if self.args.eval_bleu:
 
             def sum_logs(key):
-                return sum(log.get(key, 0) for log in logging_outputs)
+              import torch
+              result = sum(log.get(key, 0) for log in logging_outputs)
+              if torch.is_tensor(result):
+                result = result.cpu()
+              return result
 
             counts, totals = [], []
             for i in range(EVAL_BLEU_ORDER):
-- 
2.31.1

